Adds new endpoint for paginating games filtered by user and sorted by creation date

TolkienFan17 opened the pull request:
This PR adds a new endpoint for paginating games, which are filtered by the calling authenticated user and are sorted by creation date with the most recent first. Noteworthily, I employ the specification with strong coupling to Ktorm to move the domain logic out of the repository, but I minimize the coupling by offering a factory.

## Definition of Done

1. The User Story has and passes all tests with consistent, reproducible results. (80% coverage)
    - There are a few end-to-end tests, several integration tests, and some unit tests. See CI below for results.

1. The User Story conforms to the style guide.
    - CI automatically runs a linter, which serves as a basic definition of conformity of styling. See CI for results.

1. The User Story contains a basic level of documentation (KDocs, comments, clear variable names, etc.)
    - All new or modified classes in the `main` source set now have up-to-date KDocs. I have added comments as I deemed necessary, and I will add more if necessary based on code review.

1. The User Story passes the following checklist from a reviewer:

    - File structure is maintained (e.g., a component goes in the Components folder)
        - I continue to use the massive service structure currently being used, and I exercised my best judgment in adding new classes.

    - Confirm that the tests are actually testing the code
        - I attempted to make the tests focus as much as possible on the relevant interactions at each layer.

    - Code contains comments that provide a basic understanding of the code
        - Comments were and will be added as necessary.
    
## Acceptance Criteria

1. Endpoint exists which accepts a size and an optional offset parameter.

    - See screenshots
        
<img width="354" alt="image" src="https://user-images.githubusercontent.com/66689498/232890453-9e3f63d2-43d8-4385-b841-54795f02b550.png">
<img width="277" alt="image" src="https://user-images.githubusercontent.com/66689498/232890657-62242718-40fc-4529-ba73-32a446f0bb2c.png">


1. Endpoint does not permit a size greater than 100 by returning 400.

    - See screenshots
    
<img width="439" alt="image" src="https://user-images.githubusercontent.com/66689498/232890795-e6bd0bf5-a5ba-4632-8fa3-6f56d0863d56.png">

1. Endpoint does not permit an offset beyond the number of valid games by returning 404.

    - See screenshots
    
<img width="318" alt="image" src="https://user-images.githubusercontent.com/66689498/232890897-43a1a6b4-4670-45f1-80f1-8e82a9b1fde6.png">

1. The games are returned with the most recent first based on the authenticated user calling the endpoint

    - The user is obtained from the authentication principal
    
<img width="228" alt="image" src="https://user-images.githubusercontent.com/66689498/232891193-2313748d-ab1d-457a-9d01-62090fd58b27.png">

1. All classes, methods, and related code have relevant KDocs.

    - All new or modified classes in the `main` source set now have up-to-date KDocs.

1. The new endpoint has documentation written in Github.

    - I wrote documentation in the `docs` folder, which has been linked to in the `README.md`

1. End to end tests are written ensuring the overall flow of the endpoint.

    - I wrote two happy path tests for when the number of results are less than the number requested and for when it is able to fully retrieve results with an offset. I wrote a test validating offset and size, and I wrote a test to confirm that offset is optional.

1. Integration tests are written for all necessary database logic.

    - I wrote tests for the list games method that verify that specifications get called in order and that the database is successfully queried. I also wrote tests extensively testing each specification used.

1. Other unit and integration tests are written as necessary to verify business logic.

    - I wrote tests for the domain to verify that the domain checks for bad offset and for an empty result set. I also wrote a test verifying that the domain service correctly orchestrates the calls, and I wrote some basic endpoint unit tests.

TolkienFan17 committed:
Adds sequences extension property for games

TolkienFan17 committed:
Updates game dto current player entity to be nullable

TolkienFan17 committed:
Updates gamedto to reference user as a foreign key instead of its id

TolkienFan17 committed:
Adds new method to repository to retrieve a user's most recent games with pagination support

TolkienFan17 committed:
Adds integration tests for list games method in the repository

TolkienFan17 committed:
Adds factory for specifications to reduce coupling

TolkienFan17 committed:
Adds specifications for listing games

TolkienFan17 committed:
Refactors game repository to use provided specifications in listing

TolkienFan17 committed:
Refactors tests to reflect repository change, and changes test entity ids to avoid conflict

TolkienFan17 committed:
Implements domain logic for listing games

TolkienFan17 committed:
Adds tests for the specifications for listing games

TolkienFan17 committed:
Adds spec factory mock to make tests pass again

TolkienFan17 committed:
Implements the list games endpoint with desired functionality

TolkienFan17 committed:
Adds initial end-to-end tests for the new list games endpoint

TolkienFan17 committed:
Forces entities with ids generated by the database rather than tests to have a high value to fix flaky tests

TolkienFan17 committed:
Expands end-to-end test coverage for the list games endpoint

TolkienFan17 committed:
Updates list game endpoint to return 404 for bad offset

This was specified as an acceptance criteria in the story, and the end-to-end test caught this bug.

TolkienFan17 committed:
Adds extensive documentation to the codebase for listing games and specification pattern

TolkienFan17 committed:
Adds external documentation for the list games endpoint

TolkienFan17 committed:
Adds unit tests for listing games in the domain layer

TolkienFan17 committed:
Adds unit tests for the list games endpoint

Kirbitz commented on documentation/list_games.md:
````diff
@@ -0,0 +1,126 @@
+
+```javascript
+{
+  method: 'GET',
+  url: '/api/game/list',
+  responseType: 'json'
+}
````
If data is being sent in the request then it should be specified here along with headers

Kirbitz commented on src/integrations/kotlin/nomic/api/ListGamesTests.kt:
```diff
@@ -0,0 +1,110 @@
+
+        Assertions.assertThat(entity.body?.success).isTrue
+        Assertions.assertThat(entity.body?.status).isEqualTo(HttpStatus.OK)
+        Assertions.assertThat(entity.body?.data).isNotNull
+
+        val data = entity.body?.data!!
+        Assertions.assertThat(data.size).isEqualTo(4)
```
Is the first assertion necessary?
Why not just do
Assertions.assertThat(entity.body?.data!!.size).isEqualTo(4)

TolkienFan17 replied:
I suppose these are just some defensive programming habits of verifying that data is not null, especially from a black box perspective. Otherwise there could be a null pointer exception or whatever instead of a more descriptive "assertion failed expected blah blah blah to be not null but was null on line 31"

Kirbitz commented on src/integrations/kotlin/nomic/api/ListGamesTests.kt:
```diff
@@ -0,0 +1,110 @@
+
+        Assertions.assertThat(entity.body?.success).isTrue
+        Assertions.assertThat(entity.body?.status).isEqualTo(HttpStatus.OK)
+        Assertions.assertThat(entity.body?.data).isNotNull
+
+        val data = entity.body?.data!!
+        Assertions.assertThat(data.size).isEqualTo(2)
```
Same idea as previous comment
Is the first assertion necessary?
Why not just do
Assertions.assertThat(entity.body?.data!!.size).isEqualTo(2)

Kirbitz commented on documentation/list_games.md:
```diff
@@ -0,0 +1,126 @@
+
+- **Method:**
+
+  `GET`
+
+- **URL Params:**
+
+  None
+
+- **Data Params:**
+
+  **Required:** Size
+
+  `size: [unsigned int]` - Must be between 1 and 100
+
+  50
+
+  **Optional:** Offset - how many games to skip
+
+  `offset: [unsigned int]`
+
+  0
```
This either should be changed into a POST request or the body params should be moved to the URI

Kirbitz commented on src/integrations/kotlin/nomic/data/repositories/games/FilterGamesByUserSpecificationTest.kt:
```diff
@@ -0,0 +1,53 @@
+        private val user1: EndUser = EndUser(1610, "ListGamesUser2")
+        private val user2: EndUser = EndUser(1620, "ListGamesUser3")
+
+        @JvmStatic
+        fun test_filterByUser_valid(): Stream<EndUser> {
+            return Stream.of(user1, user2)
+        }
+
+        @JvmStatic
+        fun test_filterByUser_invalid(): Stream<EndUser> {
+            return Stream.of(
+                EndUser(-10, "BadUser1"),
+                EndUser(1619, "BadUser2")
+            )
+        }
+    }
```
What is this doing?  Is it testing the validity of the EndUser model?  If so then why is it here?

Kirbitz commented on src/integrations/kotlin/nomic/data/repositories/games/GameRepositoryTest.kt:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    @Test
+    fun test_listGames_callsSpecificationsInOrder() {
+        val repo = GameRepository(db)
+        val specUser = IGameSpecification { games -> games.filter { it.userId eq 1601 } }
+
+        var hasFirstSpecRun = false
+
+        val firstSpec = IGameSpecification {
+                games ->
+            hasFirstSpecRun = true
+            games
+        }
+
+        val secondSpec = IGameSpecification {
+                games ->
+            Assertions.assertThat(hasFirstSpecRun).isTrue
+            games
+        }
+
+        val games = repo.listGames(specUser, firstSpec, secondSpec)
+    }
```
Gonna need an explanation on this test.
What is is doing? 
What is the thing you are testing for?

TolkienFan17 replied:
This black-box test is ensuring that the repository function calls the specifications passed to it in the order that they were passed. For example, in the current use case for repository list games, if you sort before paginating, you may not have the most recent game overall, so it's important for the repository layer to apply specifications in the order passed.

Kirbitz commented on src/integrations/kotlin/nomic/data/repositories/games/GameRepositoryTest.kt:
```diff
@@ -0,0 +1,107 @@
+
+    @ParameterizedTest
+    @ValueSource(ints = [1, 2, 4, 8, 16])
+    fun test_listGames_callsAllSpecifications(specificationCount: Int) {
+        val repo = GameRepository(db)
+        val mockSpecifications = mutableListOf<IGameSpecification>()
+        var totalSpecificationCalls = 0
+
+        for (i in 1..specificationCount) {
+            mockSpecifications.add(
+                mock<IGameSpecification> {
+                    on(it.apply(any<EntitySequence<GameDTO, Games>>())) doAnswer {
+                            answer ->
+                        totalSpecificationCalls++
+                        answer.getArgument(0)
+                    }
+                }
+            )
+        }
+
+        // Filter for the performance of the test so that the repo isn't processing a bunch of entities
+        val filterByUser = IGameSpecification { games -> games.filter { it.userId eq 1601 } }
+
+        repo.listGames(filterByUser, *mockSpecifications.toTypedArray())
+
+        Assertions.assertThat(totalSpecificationCalls).isEqualTo(specificationCount)
+    }
```
Similar Idea as before
Gonna need an explanation on this test.
What is is doing?
What is the thing you are testing for?
Why are you testing this?

TolkienFan17 replied:
This is also a black box test to ensure that it calls all the specifications provided to it and doesn't skip any for whatever reason (e.g. bad for loop)

Kirbitz commented on src/integrations/kotlin/nomic/data/repositories/games/PaginateGamesSpecificationTest.kt:
```diff
@@ -0,0 +1,98 @@
+            (1600 lessEq game.gameId) and (game.gameId less 1700)
+        }
+
+        @JvmStatic
+        fun test_paginate_badOffset(): Stream<Int> {
+            return Stream.of(15, 20)
+        }
+
+        @JvmStatic
+        fun test_paginate_validOffset(): Stream<IntArray> {
+            return Stream.of(
+                intArrayOf(1602, 1603, 1611, 1612, 1613, 1614, 1621, 1622),
+                intArrayOf(1602, 1603, 1611, 1612, 1613, 1614),
+                intArrayOf(1602, 1603)
+            )
+        }
+
+        @JvmStatic
+        fun test_paginate_validSize(): Stream<Int> {
+            return Stream.of(0, 1, 2, 5, 7)
+        }
+
+        @JvmStatic
+        fun test_paginate_invalidSize_candidatesExceedSize(): Stream<Arguments> {
+            return Stream.of(
+                Arguments.of(10, 7),
+                Arguments.of(13, 7)
+            )
+        }
```
I guess this works, but readability wise unless someone knows what ParameterizedTest and MethodSource are doing these tests are very confusing

TolkienFan17 replied:
I rewrote and rearranged the parameterized tests with some comments so that it's hopefully clearer, so see what you think about the changes.

Kirbitz commented on src/main/kotlin/nomic/api/GamesEndpoint.kt:
```diff
@@ -34,6 +38,54 @@ class GamesEndpoint(val gameDomain: GameDomain) {
+     * @param[user] The user whose games will be listed, injected by the authentication framework.
+     * @return A Spring entity representing the response that gets serialized into JSON
+     */
+    @GetMapping("list")
+    fun listGames(
+        @RequestBody input: ListGamesApiRequestModel,
+        @AuthenticationPrincipal authUser: EndUser
```
Similar idea to the documentation either change the request type to post or modify the request so it does not take in a body

Kirbitz commented on src/main/kotlin/nomic/api/GamesEndpoint.kt:
```diff
@@ -34,6 +38,54 @@ class GamesEndpoint(val gameDomain: GameDomain) {
+        @AuthenticationPrincipal authUser: EndUser
+    ): ResponseEntity<*> {
+        if (0U >= input.size || 100U < input.size) {
+            return ResponseEntity(
+                ResponseFormat(
+                    false,
+                    HttpStatus.BAD_REQUEST,
+                    "You must request at least 1 game and at most 100 games."
+                ),
+                HttpStatus.BAD_REQUEST
+            )
```
Is there a reason why an InvalidArgumentException can't be thrown to let the globalExceptionHandler return the response?

Kirbitz commented on src/main/kotlin/nomic/api/GamesEndpoint.kt:
```diff
@@ -34,6 +38,54 @@ class GamesEndpoint(val gameDomain: GameDomain) {
+        val games = gameDomain.listGames(authUser, input.size, input.offset)
+
+        if (games.isEmpty) {
+            return ResponseEntity(
+                ResponseFormat(
+                    false,
+                    HttpStatus.NOT_FOUND,
+                    "Invalid offset"
+                ),
+                HttpStatus.NOT_FOUND
+            )
```
Similar idea is there not a way to let the globalExceptionHandler solve this problem instead of having the code exist twice?

Kirbitz commented on src/main/kotlin/nomic/api/models/ListGamesApiRequestModel.kt:
```diff
@@ -0,0 +1,14 @@
+ * @param[offset] How much the batch should be offset. This is used for paginating
+ * games, and the default to 0.
+ */
+data class ListGamesApiRequestModel(
+    val size: UInt = 100U,
+    val offset: UInt = 0U
+)
```
Honestly, I wish you had discussed with the team whether pagination or size handling was something that needed to be considered in this user story as this could have saved you a lot of unnecessary work.

Kirbitz commented on src/main/kotlin/nomic/data/IDatabaseSpecification.kt:
```diff
@@ -0,0 +1,23 @@
+ * @param[E] The Ktorm entity being operated on
+ * @param[T] The Ktorm table entity of [E]
+ */
+fun interface IDatabaseSpecification<E, T> : ISpecification where E : Any, T : BaseTable<E> {
+    fun apply(sequence: EntitySequence<E, T>): EntitySequence<E, T>
+}
```
Honestly, I wish you had discussed with the team whether database specification abstraction was something that needed to be considered in this user story as this could have saved you a lot of unnecessary work.

Kirbitz commented on src/main/kotlin/nomic/data/dtos/GameDTO.kt:
```diff
@@ -19,16 +21,16 @@ import java.time.LocalDate
     val createDate: LocalDate
-    val currentPlayer: Int
-    val userId: Int
+    val currentPlayer: Int?
+    val user: UserDTO
```
Honestly, I wish you had discussed with the team whether modifying DTOs to utilize Ktorm's entity sequences was something that needed to be considered in this user story as this could have saved you a lot of unnecessary work.

Kirbitz commented on src/main/kotlin/nomic/data/repositories/games/FilterGamesByUserSpecification.kt:
```diff
@@ -0,0 +1,20 @@
+ *
+ * @see[user] The user to filter the games by;
+ */
+class FilterGamesByUserSpecification(private val user: EndUser) : IGameSpecification {
+    override fun apply(sequence: EntitySequence<GameDTO, Games>): EntitySequence<GameDTO, Games> {
+        return sequence.filter { user.id eq Games.userId }
+    }
+}
```
Still confused on the actual purpose of this it would be nice to have a through explanation on the actual issue/problem you are trying to solve with this file.

TolkienFan17 replied:
This file in particular, or the specification pattern in general?

Kirbitz commented on src/main/kotlin/nomic/data/repositories/games/GameRepository.kt:
```diff
@@ -31,6 +33,24 @@ class GameRepository(private val db: Database) : IGameRepository {
+    override fun listGames(vararg specifications: IGameSpecification): List<GameModel> {
+        var sequence = db.games
+
+        for (specification in specifications) {
+            sequence = specification.apply(sequence)
+        }
```
I don't believe this is as readable as you think it is.  There is a lot of addition logic that has to be held in my head to understand what exactly is happening with this piece of code.

TolkienFan17 replied:
The idea is at the repository layer you are more interested in the hows than the whats or whys: you are focused on retrieving data from the database, so you aren't interested in the complexities of the domain. It helps this function focused on data retrieval and mapping (which those are two things arguably too many for this layer, but that's a whole other mater), while the domain focuses on domain things and gets data from the database from a very high level of abstraction, steering clear of details.

Its advantages would probably become more and more apparent if we had more use cases for listing some subset of games from the database, but even here there are some advantages, like putting the domain back in control of business decisions, and easier testing.

Kirbitz commented on src/main/kotlin/nomic/data/repositories/games/GameSpecificationFactory.kt:
```diff
@@ -0,0 +1,20 @@
+import org.springframework.stereotype.Service
+
+@Service
+class GameSpecificationFactory : IGameSpecificationFactory {
+    override fun filterByUser(user: EndUser): IGameSpecification {
+        return FilterGamesByUserSpecification(user)
+    }
+
+    override fun paginate(size: UInt, offset: UInt): IGameSpecification {
+        return PaginateGamesSpecification(size, offset)
+    }
+
+    override fun sortByMostRecentFirst(): IGameSpecification {
+        return SortGamesByMostRecentFirstSpecification()
+    }
+}
```
Honestly, I wish you had discussed with the team whether a gameSpecificationFactory abstraction was something that needed to be considered in this user story as this could have saved you a lot of unnecessary work.

Kirbitz commented on src/main/kotlin/nomic/data/repositories/games/PaginateGamesSpecification.kt:
```diff
@@ -0,0 +1,20 @@
+ * @see[size] The max number of games
+ * @see[offset] How many games should be skipped before listing games
+ */
+class PaginateGamesSpecification(private val size: UInt, private val offset: UInt) : IGameSpecification {
+    override fun apply(sequence: EntitySequence<GameDTO, Games>): EntitySequence<GameDTO, Games> {
+        return sequence.drop(offset.toInt()).take(size.toInt())
+    }
+}
```
Honestly, I wish you had discussed with the team whether pagination was something that needed to be considered in this user story as this could have saved you a lot of unnecessary work.

Kirbitz commented on src/main/kotlin/nomic/data/repositories/games/SortGamesByMostRecentFirstSpecification.kt:
```diff
@@ -0,0 +1,16 @@
+ * This specification sorts the games in the query by their creation dates in descending order.
+ * This results in the newest games being foremost in the query
+ */
+class SortGamesByMostRecentFirstSpecification : IGameSpecification {
+    override fun apply(sequence: EntitySequence<GameDTO, Games>): EntitySequence<GameDTO, Games> {
+        return sequence.sortedByDescending { it.createDate }
+    }
+}
```
Honestly, I wish you had discussed with the team whether sorting and what to be sorted by was something that needed to be considered in this user story as this could have saved you a lot of unnecessary work.

Kirbitz commented on src/main/kotlin/nomic/domain/games/GameDomain.kt:
```diff
@@ -25,6 +30,22 @@ class GameDomain(private val gameRepository: GameRepository) : IGameDomain {
+    override fun listGames(user: EndUser, size: UInt, offset: UInt): Optional<List<GameModel>> {
+        // Max size and offset conditions are endpoint logic rather than business logic.
+
+        val games = gameRepository.listGames(
+            specFactory.filterByUser(user),
+            specFactory.sortByMostRecentFirst(),
+            specFactory.paginate(size, offset)
+        )
```
Sure, but again I do not believe this is as readable as you think it is

TolkienFan17 replied:
Ideally, I could just create a specification directly, instead of offloading to the factory. Unfortunately there just isn't a quick and easy way to convert an expression dealing with domain entities into a Ktorm-friendly expression, so introducing reduces the coupling between Ktorm and the domain service.

Otherwise, this encapsulates the how of filtering, sorting, and paginating so that you can read this function from the perspective of the domain layer.

Kirbitz commented on src/main/kotlin/nomic/domain/games/GameDomain.kt:
```diff
@@ -25,6 +30,22 @@ class GameDomain(private val gameRepository: GameRepository) : IGameDomain {
+            specFactory.paginate(size, offset)
+        )
+
+        if (offset > 0u && games.isEmpty()) {
+            return Optional.empty()
+        }
```
Is there a reason that option is needed here?  The system could just as easily throw an exception for the globalExceptionHandler to manage.  Rather than offloading that work onto the endpoint

TolkienFan17 replied:
My thinking is that this is a predictable and expectable error that should be handled explicitly rather than implicitly. By using Optional as a poor man's result object, the caller is forced to consider the error as well. Otherwise, using an exception becomes a goto where you have to know that it's in globalExceptionHandler. I'd rather keep exceptions for unexpected errors like "can't connect to database" and use result objects to explicitly state expectable error.

Here are some articles that kinda explain my line of thinking:
https://enterprisecraftsmanship.com/posts/error-handling-exception-or-result/
https://medium.com/codex/the-error-of-exceptions-3aed074c40dc
https://patrickroza.com/blog/when-to-return-a-result-error-and-when-to-throw-an-exception/
https://www.rubypigeon.com/posts/result-objects-errors-without-exceptions/

(Obviously, this applies to some of your other PR comments)

TolkienFan17 replied:
A better solution would be a dedicated and clearer result object, which I will be more than willing to migrate the codebase to using in a separate PR after this one is closed (in spare time of course, unless for some reason the team prioritizes it higher)

Kirbitz reviewed (commented):
Here is your review

TolkienFan17 committed:
Refactors endpoint to accept parameters as url parameters rather than body parameters

The semantics of GET imply path or url parameters rather than body parameters, so this fixes it to be more semantically correct

TolkienFan17 committed:
Removes unused variables in game repo test

TolkienFan17 committed:
Refactors parameterized tests to add clarity in how they work

TolkienFan17 committed:
Refactors the api endpoint to use a new error parser to prevent code duplication

Kirbitz reviewed (approved):
is good

Tearnen reviewed (approved):
Looks good, all previous points brought up by others were addressed.
